From cf2d446dae9d566404bda8b8c03b606bd7b132e7 Mon Sep 17 00:00:00 2001
From: Jose Dapena Paz <jdapena@igalia.com>
Date: Thu, 27 Jul 2023 15:14:35 +0200
Subject: GCC: do not asume __has_feature is declared in openssl hwasan check

Instead of checking directly if hwaddress_sanitizer is enabled and
Clang version is 10+, just define OPENSSL_HWASAN, making sure the
check is only done when __has_feature is available.

Bug: 819294
Change-Id: I2bbc85d9e63326df08bb7ed091d30e68c4bb61e1
---
 .../boringssl/apple-aarch64/crypto/chacha/chacha-armv8-apple.S  | 2 +-
 .../apple-aarch64/crypto/fipsmodule/sha1-armv8-apple.S          | 2 +-
 .../apple-aarch64/crypto/fipsmodule/sha256-armv8-apple.S        | 2 +-
 .../apple-aarch64/crypto/fipsmodule/sha512-armv8-apple.S        | 2 +-
 .../boringssl/linux-aarch64/crypto/chacha/chacha-armv8-linux.S  | 2 +-
 .../linux-aarch64/crypto/fipsmodule/sha1-armv8-linux.S          | 2 +-
 .../linux-aarch64/crypto/fipsmodule/sha256-armv8-linux.S        | 2 +-
 .../linux-aarch64/crypto/fipsmodule/sha512-armv8-linux.S        | 2 +-
 .../boringssl/win-aarch64/crypto/chacha/chacha-armv8-win.S      | 2 +-
 .../boringssl/win-aarch64/crypto/fipsmodule/sha1-armv8-win.S    | 2 +-
 .../boringssl/win-aarch64/crypto/fipsmodule/sha256-armv8-win.S  | 2 +-
 .../boringssl/win-aarch64/crypto/fipsmodule/sha512-armv8-win.S  | 2 +-
 12 files changed, 12 insertions(+), 12 deletions(-)

diff --git a/third_party/boringssl/apple-aarch64/crypto/chacha/chacha-armv8-apple.S b/third_party/boringssl/apple-aarch64/crypto/chacha/chacha-armv8-apple.S
index 209956bac7b12..cc74cc371cd89 100644
--- a/third_party/boringssl/apple-aarch64/crypto/chacha/chacha-armv8-apple.S
+++ b/third_party/boringssl/apple-aarch64/crypto/chacha/chacha-armv8-apple.S
@@ -28,7 +28,7 @@ Lone:
 _ChaCha20_ctr32:
 	AARCH64_VALID_CALL_TARGET
 	cbz	x2,Labort
-#if __has_feature(hwaddress_sanitizer) && __clang_major__ >= 10
+#if defined(OPENSSL_HWASAN)
 	adrp	x5,:pg_hi21_nc:_OPENSSL_armcap_P
 #else
 	adrp	x5,_OPENSSL_armcap_P@PAGE
diff --git a/third_party/boringssl/apple-aarch64/crypto/fipsmodule/sha1-armv8-apple.S b/third_party/boringssl/apple-aarch64/crypto/fipsmodule/sha1-armv8-apple.S
index a67bc3848a9ed..c9852e8592b22 100644
--- a/third_party/boringssl/apple-aarch64/crypto/fipsmodule/sha1-armv8-apple.S
+++ b/third_party/boringssl/apple-aarch64/crypto/fipsmodule/sha1-armv8-apple.S
@@ -17,7 +17,7 @@
 _sha1_block_data_order:
 	// Armv8.3-A PAuth: even though x30 is pushed to stack it is not popped later.
 	AARCH64_VALID_CALL_TARGET
-#if __has_feature(hwaddress_sanitizer) && __clang_major__ >= 10
+#if defined(OPENSSL_HWASAN)
 	adrp	x16,:pg_hi21_nc:_OPENSSL_armcap_P
 #else
 	adrp	x16,_OPENSSL_armcap_P@PAGE
diff --git a/third_party/boringssl/apple-aarch64/crypto/fipsmodule/sha256-armv8-apple.S b/third_party/boringssl/apple-aarch64/crypto/fipsmodule/sha256-armv8-apple.S
index b8c7592affb3f..feca17e53fe43 100644
--- a/third_party/boringssl/apple-aarch64/crypto/fipsmodule/sha256-armv8-apple.S
+++ b/third_party/boringssl/apple-aarch64/crypto/fipsmodule/sha256-armv8-apple.S
@@ -59,7 +59,7 @@
 _sha256_block_data_order:
 	AARCH64_VALID_CALL_TARGET
 #ifndef	__KERNEL__
-#if __has_feature(hwaddress_sanitizer) && __clang_major__ >= 10
+#if defined(OPENSSL_HWASAN)
 	adrp	x16,:pg_hi21_nc:_OPENSSL_armcap_P
 #else
 	adrp	x16,_OPENSSL_armcap_P@PAGE
diff --git a/third_party/boringssl/apple-aarch64/crypto/fipsmodule/sha512-armv8-apple.S b/third_party/boringssl/apple-aarch64/crypto/fipsmodule/sha512-armv8-apple.S
index a02ec4f801848..bd45cdff44a3d 100644
--- a/third_party/boringssl/apple-aarch64/crypto/fipsmodule/sha512-armv8-apple.S
+++ b/third_party/boringssl/apple-aarch64/crypto/fipsmodule/sha512-armv8-apple.S
@@ -59,7 +59,7 @@
 _sha512_block_data_order:
 	AARCH64_VALID_CALL_TARGET
 #ifndef	__KERNEL__
-#if __has_feature(hwaddress_sanitizer) && __clang_major__ >= 10
+#if defined(OPENSSL_HWASAN)
 	adrp	x16,:pg_hi21_nc:_OPENSSL_armcap_P
 #else
 	adrp	x16,_OPENSSL_armcap_P@PAGE
diff --git a/third_party/boringssl/linux-aarch64/crypto/chacha/chacha-armv8-linux.S b/third_party/boringssl/linux-aarch64/crypto/chacha/chacha-armv8-linux.S
index a4a1e71dfb7ff..cdbfe93e05b7e 100644
--- a/third_party/boringssl/linux-aarch64/crypto/chacha/chacha-armv8-linux.S
+++ b/third_party/boringssl/linux-aarch64/crypto/chacha/chacha-armv8-linux.S
@@ -28,7 +28,7 @@
 ChaCha20_ctr32:
 	AARCH64_VALID_CALL_TARGET
 	cbz	x2,.Labort
-#if __has_feature(hwaddress_sanitizer) && __clang_major__ >= 10
+#if defined(OPENSSL_HWASAN)
 	adrp	x5,:pg_hi21_nc:OPENSSL_armcap_P
 #else
 	adrp	x5,OPENSSL_armcap_P
diff --git a/third_party/boringssl/linux-aarch64/crypto/fipsmodule/sha1-armv8-linux.S b/third_party/boringssl/linux-aarch64/crypto/fipsmodule/sha1-armv8-linux.S
index 27a6278089c12..ac7a867b9bd17 100644
--- a/third_party/boringssl/linux-aarch64/crypto/fipsmodule/sha1-armv8-linux.S
+++ b/third_party/boringssl/linux-aarch64/crypto/fipsmodule/sha1-armv8-linux.S
@@ -17,7 +17,7 @@
 sha1_block_data_order:
 	// Armv8.3-A PAuth: even though x30 is pushed to stack it is not popped later.
 	AARCH64_VALID_CALL_TARGET
-#if __has_feature(hwaddress_sanitizer) && __clang_major__ >= 10
+#if defined(OPENSSL_HWASAN)
 	adrp	x16,:pg_hi21_nc:OPENSSL_armcap_P
 #else
 	adrp	x16,OPENSSL_armcap_P
diff --git a/third_party/boringssl/linux-aarch64/crypto/fipsmodule/sha256-armv8-linux.S b/third_party/boringssl/linux-aarch64/crypto/fipsmodule/sha256-armv8-linux.S
index 8c54eb477c2ab..7dd44ce39cb02 100644
--- a/third_party/boringssl/linux-aarch64/crypto/fipsmodule/sha256-armv8-linux.S
+++ b/third_party/boringssl/linux-aarch64/crypto/fipsmodule/sha256-armv8-linux.S
@@ -59,7 +59,7 @@
 sha256_block_data_order:
 	AARCH64_VALID_CALL_TARGET
 #ifndef	__KERNEL__
-#if __has_feature(hwaddress_sanitizer) && __clang_major__ >= 10
+#if defined(OPENSSL_HWASAN)
 	adrp	x16,:pg_hi21_nc:OPENSSL_armcap_P
 #else
 	adrp	x16,OPENSSL_armcap_P
diff --git a/third_party/boringssl/linux-aarch64/crypto/fipsmodule/sha512-armv8-linux.S b/third_party/boringssl/linux-aarch64/crypto/fipsmodule/sha512-armv8-linux.S
index f13cc7b254999..a3de944762973 100644
--- a/third_party/boringssl/linux-aarch64/crypto/fipsmodule/sha512-armv8-linux.S
+++ b/third_party/boringssl/linux-aarch64/crypto/fipsmodule/sha512-armv8-linux.S
@@ -59,7 +59,7 @@
 sha512_block_data_order:
 	AARCH64_VALID_CALL_TARGET
 #ifndef	__KERNEL__
-#if __has_feature(hwaddress_sanitizer) && __clang_major__ >= 10
+#if defined(OPENSSL_HWASAN)
 	adrp	x16,:pg_hi21_nc:OPENSSL_armcap_P
 #else
 	adrp	x16,OPENSSL_armcap_P
diff --git a/third_party/boringssl/win-aarch64/crypto/chacha/chacha-armv8-win.S b/third_party/boringssl/win-aarch64/crypto/chacha/chacha-armv8-win.S
index 95ac85247a54f..3685e99cfb9e3 100644
--- a/third_party/boringssl/win-aarch64/crypto/chacha/chacha-armv8-win.S
+++ b/third_party/boringssl/win-aarch64/crypto/chacha/chacha-armv8-win.S
@@ -30,7 +30,7 @@ Lone:
 ChaCha20_ctr32:
 	AARCH64_VALID_CALL_TARGET
 	cbz	x2,Labort
-#if __has_feature(hwaddress_sanitizer) && __clang_major__ >= 10
+#if defined(OPENSSL_HWASAN)
 	adrp	x5,:pg_hi21_nc:OPENSSL_armcap_P
 #else
 	adrp	x5,OPENSSL_armcap_P
diff --git a/third_party/boringssl/win-aarch64/crypto/fipsmodule/sha1-armv8-win.S b/third_party/boringssl/win-aarch64/crypto/fipsmodule/sha1-armv8-win.S
index 6ddc339a8c52e..7b96d2c36c616 100644
--- a/third_party/boringssl/win-aarch64/crypto/fipsmodule/sha1-armv8-win.S
+++ b/third_party/boringssl/win-aarch64/crypto/fipsmodule/sha1-armv8-win.S
@@ -19,7 +19,7 @@
 sha1_block_data_order:
 	// Armv8.3-A PAuth: even though x30 is pushed to stack it is not popped later.
 	AARCH64_VALID_CALL_TARGET
-#if __has_feature(hwaddress_sanitizer) && __clang_major__ >= 10
+#if defined(OPENSSL_HWASAN)
 	adrp	x16,:pg_hi21_nc:OPENSSL_armcap_P
 #else
 	adrp	x16,OPENSSL_armcap_P
diff --git a/third_party/boringssl/win-aarch64/crypto/fipsmodule/sha256-armv8-win.S b/third_party/boringssl/win-aarch64/crypto/fipsmodule/sha256-armv8-win.S
index 3c0c327f35fd8..b6e9bb877d9bc 100644
--- a/third_party/boringssl/win-aarch64/crypto/fipsmodule/sha256-armv8-win.S
+++ b/third_party/boringssl/win-aarch64/crypto/fipsmodule/sha256-armv8-win.S
@@ -61,7 +61,7 @@
 sha256_block_data_order:
 	AARCH64_VALID_CALL_TARGET
 #ifndef	__KERNEL__
-#if __has_feature(hwaddress_sanitizer) && __clang_major__ >= 10
+#if defined(OPENSSL_HWASAN)
 	adrp	x16,:pg_hi21_nc:OPENSSL_armcap_P
 #else
 	adrp	x16,OPENSSL_armcap_P
diff --git a/third_party/boringssl/win-aarch64/crypto/fipsmodule/sha512-armv8-win.S b/third_party/boringssl/win-aarch64/crypto/fipsmodule/sha512-armv8-win.S
index 082b649f213dd..dd128c6a14e15 100644
--- a/third_party/boringssl/win-aarch64/crypto/fipsmodule/sha512-armv8-win.S
+++ b/third_party/boringssl/win-aarch64/crypto/fipsmodule/sha512-armv8-win.S
@@ -61,7 +61,7 @@
 sha512_block_data_order:
 	AARCH64_VALID_CALL_TARGET
 #ifndef	__KERNEL__
-#if __has_feature(hwaddress_sanitizer) && __clang_major__ >= 10
+#if defined(OPENSSL_HWASAN)
 	adrp	x16,:pg_hi21_nc:OPENSSL_armcap_P
 #else
 	adrp	x16,OPENSSL_armcap_P
-- 
2.39.2

